tests/raftstore/pd.rs

```diff
@@ -82,114 +99,357 @@ impl Cluster {
+                    change_peer.set_peer(peer.clone());
+                    resp.set_change_peer(change_peer);
+                    break;
+                }
+            }
+        } else if peer_number > max_peer_number {
+            // find the first peer which not leader.
+            let pos = region.get_peers()
+                .iter()
+                .position(|x| x.get_store_id() != leader.get_store_id())
+                .unwrap();
+
+            change_peer.set_change_type(raftpb::ConfChangeType::RemoveNode);
+            change_peer.set_peer(region.get_peers()[pos].clone());
+            resp.set_change_peer(change_peer);
 
-        // origin pre-split region's end key is the same as right end key,
-        // and must exists.
-        if !self.regions.contains_key(&right_end_key) {
-            return Err(box_err!("region {:?} doesn't exist", right));
         }
 
-        if self.regions.contains_key(&left_end_key) {
-            return Err(box_err!("region {:?} has already existed", left));
-        }
+        Ok(resp)
+    }
 
-        assert!(self.region_id_keys.insert(left.get_id(), left_end_key.clone()).is_some());
-        assert!(self.region_id_keys.insert(right.get_id(), right_end_key.clone()).is_none());
+    fn region_heartbeat(&mut self,
+                        region: metapb::Region,
+                        leader: metapb::Peer)
+                        -> Result<pdpb::RegionHeartbeatResponse> {
+        try!(self.handle_heartbeat_version(region.clone()));
+        self.handle_heartbeat_conf_ver(region, leader)
+    }
+}
 
-        for peer in right.get_peers() {
-            let store = self.stores.get_mut(&peer.get_store_id()).unwrap();
-            store.region_ids.insert(right.get_id());
-        }
+fn check_stale_region(region: &metapb::Region, check_region: &metapb::Region) -> Result<()> {
+    let epoch = region.get_region_epoch();
+    let check_epoch = check_region.get_region_epoch();
+    if check_epoch.get_conf_ver() >= epoch.get_conf_ver() &&
+       check_epoch.get_version() >= epoch.get_version() {
+        return Ok(());
+    }
 
-        assert!(self.regions.insert(left_end_key, left).is_none());
-        assert!(self.regions.insert(right_end_key, right).is_some());
-        debug!("cluster.regions: {:?}", self.regions);
+    Err(box_err!("stale epoch {:?}, we are now {:?}", check_epoch, epoch))
+}
 
-        Ok(())
+fn must_same_peers(left: &metapb::Region, right: &metapb::Region) {
+    assert_eq!(left.get_peers().len(), right.get_peers().len());
+    for peer in left.get_peers() {
+        let p = find_peer(&right, peer.get_store_id()).unwrap();
+        assert_eq!(p.get_id(), peer.get_id());
     }
+}
 
-    fn get_stores(&self) -> Vec<metapb::Store> {
-        self.stores.values().map(|s| s.store.clone()).collect()
+// Left - Right, left (1, 2, 3), right (1, 2), left - right = (3)
+fn different_peers(left: &metapb::Region, right: &metapb::Region) -> Vec<metapb::Peer> {
```

Comment:
I think `setdiff_peers` would be a better name. Refer: https://en.wikipedia.org/wiki/Complement_(set_theory)#Complements_in_various_programming_languages